File: 221/js/221.js
/**
 * @param {character[][]} matrix
 * @return {number}
 */
var maximalSquare = function(matrix) {
    // make 2d dp
    let m = matrix.length;
    let n = matrix[0].length;
    let dp = new Array(m).fill(null).map(() => Array(n).fill(0));
    let maxSqu = 0;

    // State transition equation
    for (let i = 0; i < m; i++){
        for (let j = 0; j < n; j++){
            if (matrix[i][j] === "1") {
                if (i === 0 || j === 0){
                    dp[i][j] = 1;
                }else{
                    dp[i][j] = 1 + Math.min(dp[i][j-1], dp[i-1][j], dp[i-1][j-1]);
                }

                if (dp[i][j] > maxSqu) {
                    maxSqu = dp[i][j];
                }
            }
        }
    }

    return maxSqu * maxSqu;
};

File: 221/py/221.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        m = len(matrix)
        n = len(matrix[0])
        dp = [[ 0 for _ in range(n)] for _ in range(m)]
        maxSqu = 0

        for i in range(m):
            for j in range(n):
                if matrix[i][j] == "1":
                    if i == 0 or j == 0 :
                        dp[i][j] = 1
                    else :
                        dp[i][j] =  1 + min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1])
                    
                    if maxSqu < dp[i][j] :
                        maxSqu = dp[i][j]

        return maxSqu * maxSqu
